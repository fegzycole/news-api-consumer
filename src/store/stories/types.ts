export const SET_STORIES = "SET_STORIES";
export const SET_ERROR = "SET_ERROR";
export const SET_FETCHING_STORIES = "SET_FETCHING_STORIES";

export enum StoryCategories {
  TopHeadlines = "Top Headlines",
  Business = "Business",
  Entertainment = "Entertainment",
  General = "General",
  Health = "Health",
  Science = "Science",
  Sports = "Sports",
  Technology = "Technology",
}

export interface StoryData {
  source: {
    id: null | number;
    name: string;
  };
  author: string;
  title: string;
  description: string;
  url: string;
  urlToImage: string;
  publishedAt: Date | string;
  content: string;
}

export interface StoriesState {
  stories: StoryData[];
  fetchingStories: boolean;
  error?: string;
}

export interface AddStories {
  stories: StoryData[];
  clearStories: boolean;
}

export interface SetStories {
  type: typeof SET_STORIES;
  payload: AddStories;
}

export interface SetError {
  type: typeof SET_ERROR;
  payload: string;
}

export interface SetFetchingStories {
  type: typeof SET_FETCHING_STORIES;
  payload: boolean;
}

export type MoviesActionTypes =
  | SetStories
  | SetError
  | SetFetchingStories;
