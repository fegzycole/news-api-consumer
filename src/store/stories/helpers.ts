import { StoryCategories } from "./types";

export const apiKey = process.env.REACT_APP_API_KEY;
export const apiUrl = "https://newsapi.org/v2/top-headlines";

export const getRequestPath = (
  category: StoryCategories,
  pageNumber: number
): string => {
  let path = `${apiUrl}?apiKey=${apiKey}&pageSize=10&page=${pageNumber}&language=en&sortBy=publishedAt`;

  if (category !== StoryCategories.TopHeadlines) {
    path = `${path}&category=${category.toLowerCase()}`;
  }

  return path;
};

export const getResponseError = (error: any): string => {
  let serverErrResponse = error?.response?.data?.message;

  if (!serverErrResponse) {
    serverErrResponse = error.message;
  }

  return serverErrResponse;
};
