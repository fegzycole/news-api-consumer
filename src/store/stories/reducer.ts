import {
  MoviesActionTypes,
  StoriesState,
  SET_STORIES,
  SET_ERROR,
  SET_FETCHING_STORIES,
} from "./types";

const initialState: StoriesState = {
  stories: [],
  fetchingStories: false,
  error: undefined,
};

const storiesReducer = (
  state: StoriesState = initialState,
  action: MoviesActionTypes
) => {
  switch (action.type) {
    case SET_STORIES:
      return {
        ...state,
        stories: action.payload.clearStories
          ? action.payload.stories
          : state.stories.concat(action.payload.stories),
        fetchingStories: false,
      };
    case SET_ERROR:
      return {
        ...state,
        error: action.payload,
        fetchingStories: false,
      };
    case SET_FETCHING_STORIES:
      return {
        ...state,
        fetchingStories: action.payload,
      };
    default:
      return state;
  }
};

export default storiesReducer;
