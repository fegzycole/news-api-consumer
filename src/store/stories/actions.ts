import { Dispatch } from "react";
import axios from "axios";

import {
  StoryData,
  SetStories,
  SET_STORIES,
  SetError,
  SET_ERROR,
  SetFetchingStories,
  SET_FETCHING_STORIES,
  StoryCategories,
  AddStories,
} from "./types";
import { getResponseError, getRequestPath } from "./helpers";

export const setStories = (payload: AddStories): SetStories => {
  return {
    type: SET_STORIES,
    payload,
  };
};

export const setError = (payload: string): SetError => {
  return {
    type: SET_ERROR,
    payload,
  };
};

export const setFetchingStories = (payload: boolean): SetFetchingStories => {
  return {
    type: SET_FETCHING_STORIES,
    payload,
  };
};

export const getStories = (
  category: StoryCategories,
  pageNumber: number
): ((
  dispatch: Dispatch<{ type: string; payload?: unknown }>
) => Promise<void>) => {
  return async (dispatch) => {
    dispatch(setError(""));
    dispatch(setFetchingStories(true));

    const requestPath = getRequestPath(category, pageNumber);

    try {
      const request = await axios.get(requestPath);

      const responseData = request.data.articles as StoryData[];

      const stories: StoryData[] = responseData as StoryData[];

      const payload: AddStories = {
        stories,
        clearStories: pageNumber === 1,
      };

      return dispatch(setStories(payload));
    } catch (error: any) {
      const renderedError = getResponseError(error);
      return dispatch(setError(renderedError));
    }
  };
};
