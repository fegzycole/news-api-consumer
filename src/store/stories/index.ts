import * as Types from './types';
import storiesReducer from './reducer';
import * as Actions from './actions';

export { Types, storiesReducer, Actions };
