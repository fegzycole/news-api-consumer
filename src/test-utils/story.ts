import { StoryData } from "../store/stories/types";

const generateStories = () => {
  const mockStories: StoryData[] = [];

  let i = 0;

  while (i < 12) {
    mockStories.push(mockStory);

    i++;
  }

  return mockStories;
};

export const mockStory: StoryData = {
  publishedAt: "Mock Date",
  author: "Mock Author",
  urlToImage: "MockUrlToImage",
  description: "Mock Description",
  source: {
    name: "Mock Source Name",
    id: 1,
  },
  title: "Mock Title",
  url: "mockUrl",
  content: "MockContent",
};

export const mockStories: StoryData[] = generateStories();
