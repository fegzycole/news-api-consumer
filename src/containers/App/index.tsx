import { useState, useEffect, useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import { AnyAction } from "redux";
import { Report } from "notiflix/build/notiflix-report-aio";

import Loader from "../../components/Loader";
import Header from "../../components/Header";
import { RootState } from "../../store";
import { getStories } from "../../store/stories/actions";
import { StoryCategories } from "../../store/stories/types";
import Stories from "../Stories";
import PaginationBtn from "../../components/Pagination";

const PAGE_NUMBER = 1;

function App() {
  const dispatch = useDispatch();

  const { stories, fetchingStories, error } = useSelector(
    (state: RootState) => state.stories
  );

  const [selectedCategory, setSelectedCategory] = useState<StoryCategories>(
    StoryCategories.TopHeadlines
  );

  const [currentPageNumber, setCurrentPageNumber] =
    useState<number>(PAGE_NUMBER);

  const fetchStories = useCallback(
    async (category: StoryCategories, page: number) => {
      await dispatch(getStories(category, page) as unknown as AnyAction);
    },
    [dispatch]
  );

  const handleClick = (val: StoryCategories) => {
    setSelectedCategory(val);
    setCurrentPageNumber(PAGE_NUMBER);
  };

  const increasePageCount = () => {
    const newPageNumber = currentPageNumber + PAGE_NUMBER;
    setCurrentPageNumber(newPageNumber);
  };

  useEffect(() => {
    fetchStories(selectedCategory, currentPageNumber);
  }, [fetchStories, selectedCategory, currentPageNumber]);

  useEffect(() => {
    if (error) {
      Report.failure("Error", error, "Close");
    }
  }, [error]);

  const hideLoader = !fetchingStories && stories && stories.length;

  return (
    <div className="Container">
      <Header selectedCategory={selectedCategory} handleClick={handleClick} />
      {hideLoader ? (
        <main>
          <Stories
            stories={stories}
            topStory={stories[0]}
            selectedCategory={selectedCategory}
          />
          <PaginationBtn handleClick={increasePageCount} />
        </main>
      ) : (
        <Loader />
      )}
    </div>
  );
}

export default App;
