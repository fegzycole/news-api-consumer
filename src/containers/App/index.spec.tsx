import { render } from "@testing-library/react";
import { useDispatch, useSelector } from "react-redux";

import { StoryData } from "../../store/stories/types";
import { mockStory } from "../../test-utils/story";
import App from "./index";

const generateStories = () => {
  const mockStories: StoryData[] = [];

  let i = 0;

  while (i < 12) {
    mockStories.push(mockStory);

    i++;
  }

  return mockStories;
};

const mockUseDispatch = useDispatch as jest.Mock<any>;
const mockUseSelector = useSelector as jest.Mock<any>;

const mockDispatch = jest.fn();

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock("notiflix/build/notiflix-report-aio", () => ({
  Report: {
    failure: jest.fn(),
  },
}));

jest.mock("../../components/Loader", () => ({
  default: () => <div className="Loader" />,
  __esModule: true,
  Loader: true,
}));

jest.mock("../../components/Header", () => ({
  default: () => <div className="Header" />,
  __esModule: true,
  Header: true,
}));

jest.mock("../Stories", () => ({
  default: () => <div className="Stories" />,
  __esModule: true,
  Stories: true,
}));

jest.mock("../../components/Pagination", () => ({
  default: () => <div className="Pagination" />,
  __esModule: true,
  Pagination: true,
}));

jest.mock("../../store/stories/actions", () => ({
  getStories: jest.fn(),
}));

describe("<App />", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    mockUseDispatch.mockImplementation(() => mockDispatch);
    mockUseSelector.mockImplementation((selectorFn: any) =>
      selectorFn({
        stories: {
          stories: generateStories(),
          topStory: mockStory,
          error: "Mock Error",
          fetchingStories: false,
        },
      })
    );
  });

  it("should match the App component snapshot", () => {
    const view = render(<App />);
    expect(view.container).toMatchSnapshot();
  });
});
