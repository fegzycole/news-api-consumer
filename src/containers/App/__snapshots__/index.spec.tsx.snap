// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match the App component snapshot 1`] = `
<div>
  <div
    class="Container"
  >
    <div
      class="Header"
    />
    <main>
      <div
        class="Stories"
      />
      <div
        class="Pagination"
      />
    </main>
  </div>
</div>
`;
